yjit/src/codegen.rs

```diff
@@ -328,74 +313,80 @@ fn jit_prepare_routine_call(jit: &mut JITState, ctx: &mut Context, cb: &mut Code
+/// Verify the ctx's types and mappings against the compile-time stack, self,
+/// and locals.
+fn verify_ctx(jit: &JITState, ctx: &Context)
 {
-    // Only able to check types when at current insn
-    RUBY_ASSERT(jit_at_current_insn(jit));
-
-    VALUE self_val = jit_peek_at_self(jit, ctx);
-    if (type_diff(yjit_type_of_value(self_val), ctx->self_type) == INT_MAX) {
-        rb_bug("verify_ctx: ctx type (%s) incompatible with actual value of self: %s", yjit_type_name(ctx->self_type), rb_obj_info(self_val));
+    fn obj_info_str<'a>(val: VALUE) -> &'a str {
+        unsafe { CStr::from_ptr(rb_obj_info(val)).to_str().unwrap() }
     }
 
-    for (int i = 0; i < ctx->stack_size && i < MAX_TEMP_TYPES; i++) {
-        temp_type_mapping_t learned = ctx.get_opnd_mapping(StackOpnd(i));
-        VALUE val = jit_peek_at_stack(jit, ctx, i);
-        val_type_t detected = yjit_type_of_value(val);
+    // Only able to check types when at current insn
+    assert!(jit_at_current_insn(jit));
 
-        if (learned.mapping.kind == TEMP_SELF) {
-            if (self_val != val) {
-                rb_bug("verify_ctx: stack value was mapped to self, but values did not match\n"
-                        "  stack: %s\n"
-                        "  self: %s",
-                        rb_obj_info(val),
-                        rb_obj_info(self_val));
-            }
-        }
+    let self_val = jit_peek_at_self(jit, ctx);
+    let self_val_type = Type::from(self_val);
 
-        if (learned.mapping.kind == TEMP_LOCAL) {
-            int local_idx = learned.mapping.idx;
-            VALUE local_val = jit_peek_at_local(jit, ctx, local_idx);
-            if (local_val != val) {
-                rb_bug("verify_ctx: stack value was mapped to local, but values did not match\n"
-                        "  stack: %s\n"
-                        "  local %i: %s",
-                        rb_obj_info(val),
+    // Verify self operand type
+    if self_val_type.diff(ctx.get_opnd_type(SelfOpnd)) == usize::MAX {
+        panic!(
+            "verify_ctx: ctx self type ({:?}) incompatible with actual value of self {}",
+            ctx.get_opnd_type(SelfOpnd),
+            obj_info_str(self_val)
+        );
+    }
+
+    // Verify stack operand types
+    let top_idx = cmp::min(ctx.get_stack_size(), MAX_TEMP_TYPES as u16);
+    for i in 0..top_idx {
+        let (learned_mapping, learned_type) = ctx.get_opnd_mapping(StackOpnd(i));
+        let stack_val = jit_peek_at_stack(jit, ctx, i as isize);
+        let val_type = Type::from(stack_val);
+
+        match learned_mapping {
+            TempMapping::MapToSelf => {
+                if self_val != stack_val {
+                    panic!(
+                        "verify_ctx: stack value was mapped to self, but values did not match!\n  stack: {}\n  self: {}",
+                        obj_info_str(stack_val),
+                        obj_info_str(self_val)
+                    );
+                }
+            },
```

Comment:
I mostly ask because asserts on locals were what I saw failing. If we changed how they worked, then that's why they're failing and that makes sense.

If they're just failing because they're wrong, then removing the test seems like a worse idea.